rustls/tests/unbuffered.rs

```diff
@@ -0,0 +1,204 @@
+#![cfg(any(feature = "ring", feature = "aws_lc_rs"))]
+use std::sync::Arc;
+
+use rustls::client::{ClientConnectionData, UnbufferedClientConnection};
+use rustls::server::{ServerConnectionData, UnbufferedServerConnection};
+use rustls::unbuffered::{ConnectionState, UnbufferedConnectionCommon, UnbufferedStatus};
+
+use crate::common::*;
+
+mod common;
+
+#[test]
+fn handshake() {
+    for version in rustls::ALL_VERSIONS {
+        let server_config = make_server_config(KeyType::Rsa);
+        let client_config = make_client_config_with_versions(KeyType::Rsa, &[version]);
+
+        let mut client =
+            UnbufferedClientConnection::new(Arc::new(client_config), server_name("localhost"))
+                .unwrap();
+        let mut server = UnbufferedServerConnection::new(Arc::new(server_config)).unwrap();
+        let mut buffers = BothBuffers::default();
+
+        let mut count = 0;
+        let mut client_handshake_done = false;
+        let mut server_handshake_done = false;
+        while !client_handshake_done || !server_handshake_done {
+            match advance_client(&mut client, &mut buffers.client, NO_ACTIONS) {
+                State::EncodedTlsData => {}
+                State::MustTransmitTlsData => buffers.client_send(),
+                State::NeedsMoreTlsData => buffers.server_send(),
+                State::TrafficTransit => client_handshake_done = true,
+            }
+
+            match advance_server(&mut server, &mut buffers.server, NO_ACTIONS) {
+                State::EncodedTlsData => {}
+                State::MustTransmitTlsData => buffers.server_send(),
+                State::NeedsMoreTlsData => buffers.client_send(),
+                State::TrafficTransit => server_handshake_done = true,
+            }
+
+            count += 1;
+
+            assert!(count <= 100, "handshake {version:?} was not completed");
+        }
+    }
+}
+
+#[derive(Debug)]
+enum State {
+    EncodedTlsData,
+    NeedsMoreTlsData,
+    TrafficTransit,
+    MustTransmitTlsData,
+}
+
+const NO_ACTIONS: Actions = Actions {};
+
+#[derive(Clone, Copy, Debug)]
+struct Actions {}
```

Comment:
What's the purpose of the `Actions` struct? In this revision it's always `NO_ACTIONS` at the call-sites, and it's threaded through to `handle_state` where it's an ignored parameter. If a later commit turns this into something functional I think it would be less confusing to introduce it then as opposed to making it a no-op here.